# Knihovny
Python obsahuje velké množství knihoven, které jsou předinstalované a můžeme je používat. Některé z nich jsou již využity v předchozích příkladech. Nyní si ukážeme, jak je můžeme importovat a používat.

## Importování knihovny
Knihovnu můžeme importovat pomocí příkazu `import`. Následně můžeme používat všechny funkce, které knihovna obsahuje. Například knihovna `math` obsahuje funkci `sqrt`, která počítá odmocninu. Knihovnu `math` můžeme importovat pomocí příkazu `import math`. Následně můžeme používat funkci `sqrt` pomocí příkazu `math.sqrt`.

In [1]:
import math

cislo = 25
odmocnina = math.sqrt(cislo)

print(odmocnina)

5.0


Výše ukázaný postup má jednu nevýhodu. Pokud chceme používat funkci `sqrt`, musíme vždy uvádět jméno knihovny. To může být nepříjemné, pokud chceme používat funkci často. Proto existuje možnost importovat pouze funkci `sqrt` z knihovny `math`. Toho docílíme pomocí příkazu `from math import sqrt`. Následně můžeme používat funkci `sqrt` přímo.

In [2]:
from math import sqrt

cislo = 25
odmocnina = sqrt(cislo)

print(odmocnina)

5.0


Nevýhodou tohoto způsobu je, že pokud chceme používat více funkcí z knihovny, musíme je všechny importovat zvlášť. To může být nepříjemné, pokud chceme používat velké množství funkcí. Proto existuje možnost importovat všechny funkce z knihovny pomocí příkazu `from math import *`. Následně můžeme používat všechny funkce z knihovny přímo.

In [3]:
from math import *

cislo = 25
odmocnina = sqrt(cislo)

print(odmocnina)

5.0


Importovat všechny funkce z knihovny není doporučené. Pokud totiž importujeme více knihoven, může se stát, že se v nich budou nacházet funkce se stejným názvem. V takovém případě se může stát, že se nám některé funkce přepíšou a budou nefunkční. Proto je lepší importovat celou knihovnu pomocí příkazu `import math` a následně používat funkce pomocí příkazu `math.sqrt`.

Někdy názvy knihoven nebo funkcí jsou příliš dlouhé. V takovém případě můžeme knihovnu nebo funkci přejmenovat. Toho docílíme pomocí příkazu `import math as m` nebo `from math import sqrt as s`. Následně můžeme používat knihovnu nebo funkci pomocí příkazu `m.sqrt` nebo `s`.

In [4]:
import math as m

cislo = 25
odmocnina = m.sqrt(cislo)

print(odmocnina)

5.0


In [5]:
from math import sqrt as s

cislo = 25
odmocnina = s(cislo)

print(odmocnina)

5.0


## Soubor jako knihovna
Dobrou praktikou je psát kód do souborů a následně je importovat, tato praktika podporuje přenositelnost kódu. Toho docílíme pomocí příkazu `import nazev_souboru`. Následně můžeme používat všechny funkce, které se v souboru nachází. Soubor musí mít příponu `.py`. Soubor musí být ve stejné složce jako soubor, ve kterém je importován.

In [ ]:
# vypocty.py

def secti(a, b):
    return a + b

def odecti(a, b):
    return a - b

def vynasob(a, b):
    return a * b

def vydel(a, b):
    return a / b

KONSTANTA = 10

In [ ]:
# main.py

import vypocty

print(vypocty.secti(1, 2))
print(vypocty.odecti(1, 2))
print(vypocty.vynasob(1, 2))
print(vypocty.vydel(1, 2))
print(vypocty.KONSTANTA)

## __name__ == "__main__"
V souboru může být kód, který se spustí pouze v případě, že je soubor spuštěn přímo. Toho docílíme pomocí podmínky `if __name__ == "__main__":`. Tento kód se spustí pouze v případě, že je soubor spuštěn přímo. Pokud je soubor importován, kód se nespustí.

Jinými slovy, script spouštěn napřímo se vždy jmenuje `__main__`. Pokud je script importován, jmenuje se podle svého jména. Tuto vlastnost můžeme využít k tomu, abychom mohli spouštět část scriptu (např. testovací kód) pouze v případě, že je script spuštěn přímo.

In [1]:
print(__name__)

__main__


In [ ]:
# vypocty.py

def secti(a, b):
    return a + b

def odecti(a, b):
    return a - b

def vynasob(a, b):
    return a * b

def vydel(a, b):
    return a / b

if __name__ == "__main__":
    print(secti(1, 2))
    print(odecti(1, 2))
    print(vynasob(1, 2))
    print(vydel(1, 2))

In [ ]:
# main.py

import vypocty

print(vypocty.secti(1, 2))


# Cvičení
Bude následovat pár cvičení, které procvičí práci s knihovnami.

## Cvičení 1
Importujte knihovnu `math` a vypište hodnotu $\pi$.

In [6]:
from math import pi

print(pi)

3.141592653589793


### Referenční výstup
<span>Pi má hodnotu: 3.141592653589793</span>


## Cvičení 2
Importujte knihovnu `math` a vypište hodnotu $\pi$ pomocí přejmenování knihovny.

In [7]:
import math as m

print(m.pi)

3.141592653589793


### Referenční výstup
<span>Pi má hodnotu: 3.141592653589793</span>


## Cvičení 3
Importujte knihovnu `math` a vypište hodnotu $\pi$ pomocí přejmenování funkce.

In [8]:
from math import pi as p

print(p)

3.141592653589793


### Referenční výstup
<span>Pi má hodnotu: 3.141592653589793</span>

## Cvičení 4
Vytvořte soubor `soubor.py` a vypište v něm hodnotu $\pi$. Následně importujte tento soubor a vypište hodnotu $\pi$.

In [ ]:
# soubor.py

# zde přijde váš kód



In [ ]:
# main.py

# zde přijde váš kód



### Referenční výstup
<span>Pi má hodnotu: 3.141592653589793</span>

## Cvičení 5
Vytvořte soubor `soubor.py` a vypište v něm hodnotu $\pi$ pouze v případě, že je soubor spuštěn přímo. Následně importujte tento soubor a vypište hodnotu $2\pi$.

In [ ]:
# soubor.py

# zde přijde váš kód



In [ ]:
# main.py

# zde přijde váš kód

